sample_metadata/parser/generic_parser.py

```diff
@@ -399,114 +582,279 @@ async def validate_rows(
 
     async def parse_manifest(  # pylint: disable=too-many-branches
         self, file_pointer, delimiter=',', confirm=False, dry_run=False
-    ) -> Union[Dict[str, Dict], Tuple[List, List, List, List, Dict]]:
+    ) -> Dict[str, Dict]:
         """
         Parse manifest from iterable (file pointer / String.IO)
 
         Returns a dict mapping external sample ID to CPG sample ID
         """
-        proj = self.sample_metadata_project
 
-        sample_map = await self.file_pointer_to_sample_map(file_pointer, delimiter)
-        if len(sample_map) == 0:
-            raise ValueError(f'{proj}: The manifest file contains no records')
+        sample_map: Dict[str, Any] = {}
+        participant_map: Dict[str, Any] = {}
 
-        # now we can start adding!!
-        sapi = SampleApi()
-        seqapi = SequenceApi()
+        if self.has_participants(file_pointer, delimiter):
+            participant_map = await self.file_pointer_to_participant_map(
+                file_pointer, delimiter
+            )
+        else:
+            sample_map = await self.file_pointer_to_sample_map(file_pointer, delimiter)
+
+        if len(sample_map) == 0 and len(participant_map) == 0:
+            raise ValueError(
+                f'{self.sample_metadata_project}: The manifest file contains no records'
+            )
+
+        if len(sample_map) != 0:
+            return await self.parse_manifest_by_samples(
+                sample_map, confirm=confirm, dry_run=dry_run
+            )
+
+        return await self.parse_manifest_by_participants(
+            participant_map, confirm=confirm, dry_run=dry_run
+        )
```

Comment:
I think just keep this branched the whole function, you'll have a little bit of code duplication but there's no need to regroup to do the error check.

Reply:
Don't quite follow....